tests/test_portfolio_env.py

```diff
@@ -0,0 +1,165 @@
+    (3, 1, 3, (3, 3)),
+    (2, 5, 3, (2, 5, 3)),
+])
+def test_combined_setup_observation_space(make_env, datums, num_values, assets, window_size, shape):
+    fill_datums(datums, assets, num_values)
+    assert make_env(window_size).observation_space == make_box(1, 1, shape)
+
+
+def fill_datums(datums, assets, num_values, value=1):
+    for _ in range(assets):
+        series = [[value] * num_values] * 10
+        datums.add().rows(*series)
+
+
+def test_reset_returns_first_value_from_datums(make_env, datums):
+    datums.add().rows([1], [2])
+    assert_obs_eq(make_env().reset(), [1])
+
+
+@pytest.mark.parametrize('num_values, assets, window_size', [
+    (3, 1, 1),
+    (3, 2, 1),
+    (1, 3, 1),
+    (3, 1, 2),
+    (1, 1, 5),
+])
+def test_observations_follow_observation_space_shape(make_env, datums, num_values, assets, window_size):
+    fill_datums(datums, assets, num_values)
+    env = make_env(window_size)
+    assert env.reset().shape == env.observation_space.shape
+
+
+def test_reset_returns_properly_windowed_observation(make_env, datums):
+    datums.add().rows([1], [2], [3])
+    assert_obs_eq(make_env(window_size=2).reset(), [[1, 2]])
+
+
+def test_reset_combined_setup_observation(make_env, datums):
+    datums.add().rows(
+        [2, 1, 3],
+        [3, 2, 4],
+        [1, 1, 1],
+    )
+    datums.add().rows(
+        [0, -1, 1],
+        [1, 0, 2],
+        [1, 1, 1],
+    )
+    assert_obs_eq(make_env(window_size=2).reset(), [[[2, 3], [0, 1]],
+                                                    [[1, 2], [-1, 0]],
+                                                    [[3, 4], [1, 2]]])
+
+
+def test_create_returns_as_observations_when_configured(make_env, datums):
+    datums.add().rows(
+        [2, 1, 4],
+        [4, 2, 8],
+        [1, 4, 2],
+        [1, 1, 1],
+    )
+    assert_obs_eq(make_env(window_size=2, calc_returns=True).reset(), [[2, 0.25],
+                                                                       [2, 2],
+                                                                       [2, 0.25]])
+
+
+def test_stepping_the_environment_returns_next_observation(make_env, datums):
+    datums.add().rows([1], [2], [3])
+    env = make_env()
+    env.reset()
+    assert_obs_eq(unpack_obs(idle_step(env)), [2])
+    assert_obs_eq(unpack_obs(idle_step(env)), [3])
+
+
+def idle_step(env):
+    return env.step(0)
+
+
+def test_resetting_the_environment_resets_observations(make_env, datums):
+    datums.add().rows([1], [2], [3])
+    env = make_env()
+    env.reset()
+    assert_obs_eq(unpack_obs(idle_step(env)), [2])
+    env.reset()
+    assert_obs_eq(unpack_obs(idle_step(env)), [2])
+
+
+def test_environment_is_done_at_the_last_observation(make_env, datums):
+    datums.add().rows([1], [2], [3])
+    env = make_env()
+    env.reset()
+    assert not unpack_done(idle_step(env))
+    assert unpack_done(idle_step(env))
+
+
+def test_raise_an_error_when_not_enough_data_for_a_single_step(make_env, datums):
```

Comment:
might also want to check that this error is raised when stepping past a done environment